checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInferenceLogic.java

```diff
@@ -43,88 +76,620 @@ public class MustCallInferenceLogic {
+      if (typeFactory.declaredTypeHasMustCall(param)) {
+        result.add(
+            new Obligation(
+                ImmutableSet.of(new ResourceAlias(new LocalVariable(paramElement), param))));
+      }
+    }
+    return result;
+  }
+
+  /**
+   * This function returns a set of all owning fields that have been inferred in the current or
+   * previous iteration
+   *
+   * @return set of owning fields
+   */
+  private Set<VariableElement> getEnclosedOwningFields() {
+    ClassTree classTree = TreePathUtil.enclosingClass(typeFactory.getPath(enMethodTree));
+    TypeElement classElt = TreeUtils.elementFromDeclaration(classTree);
+    Set<VariableElement> enOwningFields = new HashSet<>();
+    for (Element memberElt : classElt.getEnclosedElements()) {
+      if (memberElt.getKind().isField() && typeFactory.hasOwning(memberElt)) {
+        enOwningFields.add((VariableElement) memberElt);
+      }
+    }
+    for (Element memberElt : owningFieldToECM) {
+      enOwningFields.add((VariableElement) memberElt);
+    }
+    return enOwningFields;
+  }
+
+  /**
+   * Given an index, adds an owning annotation to the parameter at the specified index
+   *
+   * @param index index of enclosing method's parameter
+   */
+  private void addOwningOnParams(int index) {
+    WholeProgramInference wpi = typeFactory.getWholeProgramInference();
+    wpi.addDeclarationAnnotationToFormalParameter(enMethodElt, index, OWNING);
+  }
+
+  /**
+   * Checks if the given node is a field and if it is a new Owning field.
+   *
+   * @param node the possible owning field
+   * @param mNode method invocation node
+   */
+  private void isOwningField(@Nullable Node node, MethodInvocationNode mNode) {
+    if (node == null) {
+      return;
+    }
+    Element nodeElt = TreeUtils.elementFromTree(node.getTree());
+    if (nodeElt == null || !nodeElt.getKind().isField()) {
+      return;
+    }
+    if (!getEnclosedOwningFields().contains(nodeElt)
+        && typeFactory.isCandidateOwningField(nodeElt)) {
+      node = NodeUtils.removeCasts(node);
+      JavaExpression target = JavaExpression.fromNode(node);
+      if (mustCallObligationSatisfied(mNode, nodeElt, target)) {
+        // This assumes that any MustCall annotation has at most one element.
+        // TODO: generalize this to MustCall annotations with more than one element.
+        owningFieldToECM.add((VariableElement) nodeElt);
+      }
+    }
+  }
+
+  /**
+   * Updates a set of obligations based on an assignment statement node. If the left-hand side of
+   * the assignment is a field that is an "enclosed owning field", adds the owning field to the
+   * method's parameters if its alias is assigned to the field. If the left-hand side of the
+   * assignment is a resource variable and the right-hand side is a must-call-close method call,
+   * adds the owning resource to the method's parameters if its alias is assigned to the resource
+   * variable. Otherwise, updates the obligations based on the assignment.
```

Comment:
How much does this have in common with `MustCallConsistencyAnalyzer#updateObligationsForAssignment`? I think they have a lot in common - looking at the code here briefly suggests that this is a modified version of that method. If so, you should factor out common elements, so that if one changes, the other will too.

Reply:
I think the only common computation is the one which updates the obligations set and I replaced it with MustCallConsistencyAnalyzer#updateObligationsForPseudoAssignment. The remaining computation is just case analysis, which should be addressed differently in the ResourceLeakInference.